flang/lib/Lower/Bridge.cpp

```diff
@@ -801,161 +798,255 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+    auto &doStmtEval = eval.getFirstNestedEvaluation();
     auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
-    assert(doStmt && "missing DO statement");
-    const auto &loopControl =
-        std::get<std::optional<Fortran::parser::LoopControl>>(doStmt->t);
-    llvm::SmallVector<IncrementLoopInfo, 1> incrementLoopInfo;
-    const Fortran::parser::ScalarLogicalExpr *whileCondition = nullptr;
+    const auto &loopControl = std::get<1>(doStmt->t);
+    auto *preheaderBlock = doStmtEval.block;
+    auto *headerBlock =
+        unstructuredContext ? doStmtEval.localBlocks[0] : nullptr;
+    auto *bodyBlock = doStmtEval.lexicalSuccessor->block;
+    auto *exitBlock = doStmtEval.parentConstruct->constructExit->block;
+    llvm::SmallVector<IncrementLoopInfo, 4> incrementLoopInfo;
+    const Fortran::parser::ScalarLogicalExpr *whileCondition{};
     bool infiniteLoop = !loopControl.has_value();
     if (infiniteLoop) {
       assert(unstructuredContext && "infinite loop must be unstructured");
-      startBlock(doStmtEval.localBlocks[0]); // header block
-    } else if ((whileCondition =
-                    std::get_if<Fortran::parser::ScalarLogicalExpr>(
-                        &loopControl->u))) {
+      startBlock(headerBlock);
+    } else if ((whileCondition = std::get_if<1>(&loopControl->u))) {
       assert(unstructuredContext && "while loop must be unstructured");
-      startBlock(doStmtEval.localBlocks[0]); // header block
-      genFIRConditionalBranch(*whileCondition, doStmtEval.lexicalSuccessor,
-                              doStmtEval.parentConstruct->constructExit);
-    } else if (const auto *bounds =
-                   std::get_if<Fortran::parser::LoopControl::Bounds>(
-                       &loopControl->u)) {
-      // "Normal" increment loop.
-      incrementLoopInfo.emplace_back(bounds->name.thing.symbol, bounds->lower,
-                                     bounds->upper, bounds->step,
-                                     genType(*bounds->name.thing.symbol));
+      startBlock(headerBlock);
+      genFIRConditionalBranch(*whileCondition, bodyBlock, exitBlock);
+    } else if (const auto *bounds = std::get_if<0>(&loopControl->u)) {
+      // Non-concurrent increment loop.
+      incrementLoopInfo.emplace_back(*bounds->name.thing.symbol, bounds->lower,
+                                     bounds->upper, bounds->step);
       if (unstructuredContext) {
-        maybeStartBlock(doStmtEval.block); // preheader block
-        incrementLoopInfo[0].headerBlock = doStmtEval.localBlocks[0];
-        incrementLoopInfo[0].bodyBlock = doStmtEval.lexicalSuccessor->block;
-        incrementLoopInfo[0].successorBlock =
-            doStmtEval.parentConstruct->constructExit->block;
+        maybeStartBlock(preheaderBlock);
+        auto &info = incrementLoopInfo.back();
+        info.hasRealControl = info.loopVariableSym.GetType()->IsNumeric(
+            Fortran::common::TypeCategory::Real);
```

Comment:
Thanks for fixing this one... If I remember right, REAL loops are verboten in 2018. I assume semantics will do something if strict 2018 conformance is specified...

Reply:
I have a cheat sheet from Gary that says REAL loops were deprecated in f90, and deleted in f95.  If someone gets serious about trying to account for deleted features under some option, then yes, that should happen upstream.